reana_server/reana_admin.py

```diff
@@ -524,25 +526,50 @@ def list_quota_resources(ctx, admin_access_token):
         "E.g. --email johndoe@example.org --email janedoe@example.org"
     ),
 )
-@click.option("--resource-name", "-r", help="Name of resource.", required=True)
+@click.option(
+    "--resource", "-r", help="Specify quota resource. e.g. cpu, disk.", required=True
+)
+@click.option("--resource-name", help="Name of resource.")
 @click.option(
     "--limit", "-l", help="New limit in canonical unit.", required=True, type=int
 )
 @admin_access_token_option
 @click.pass_context
-def set_quota_limit(ctx, admin_access_token, emails, resource_name, limit):
+def set_quota_limit(ctx, admin_access_token, emails, resource, resource_name, limit):
     """Set quota limits to the given users per resource."""
     try:
         for email in emails:
             error_msg = None
             user = _get_user_by_criteria(None, email)
-            resource = Resource.query.filter_by(name=resource_name).one_or_none()
+            try:
+                resource = Resource.query.filter(
+                    Resource.type_ == resource
+                ).one_or_none()
+            except MultipleResultsFound:
```

Comment:
What about checking `count() > 1` instead of risking an exception raise? I think it'd look cleaner and we would avoid the try-except block, session rollback, etc.

Reply:
There are some issues while querying a `Resource` by `type_` which is enum:
```
$ kubectl exec -i -t deployment/reana-server -c rest-api -- invenio shell
from reana_db.models import Resource
Resource.query.filter(Resource.type_ == "some resource").first()
# Gives: DataError: (psycopg2.errors.InvalidTextRepresentation) invalid input value for enum resourcetype: "some resource"
```
So try-except block would still be needed.
Also `resource` value is used later in the code.. with `count() > 1` we would need two queries

Reply:
It can be done with something like this:

```python
In [0]: my_type = 'cpu'

In [1]: ResourceType._member_names_
Out[1]: ['cpu', 'disk']

In [2]: my_type in ResourceType._member_names_
Out[2]: True

In [3]: ResourceType[my_type]
Out[3]: <ResourceType.cpu: 0>

In [4]: Resource.query.filter_by(type_=ResourceType[my_type]).all()
Out[4]:
[<Resource 62bcabea-f073-456e-97a7-e274ec5c1e2b>,
 <Resource 1cb18814-a556-4c54-ab4f-5d04a4f22c8c>]
```

As far as I know count queries are very cheap. But if you prefer you can fetch all the results and use `len()` instead.

Reply:
I came up with this peace of code:
```python

if resource in ResourceType._member_names_:
    resource_count = Resource.query.filter_by(type_=resource).count()
    if resource_count > 1:
        if resource_name:
            resource = Resource.query.filter_by(
                name=resource_name
            ).one_or_none()
        else:
            error_msg = (
                f"ERROR: There are more than one `{resource}` resource. "
                "Please provide resource name with `--resource-name` option to specify the exact resource."
            )
    else:
        resource = Resource.query.filter_by(type_=resource).one_or_none()
else:
    resource = None
```
which indeed avoids the try-except block and session rollback, but still looks messy IMO. Is this what you had in mind?

Reply:
Similar to that, but I think we don't 3 queries. I've been playing a little bit with the code. Here more or less is the diff of the idea I had in mind:

```diff
diff --git a/reana_server/reana_admin.py b/reana_server/reana_admin.py
index 6002de6..a093e65 100644
--- a/reana_server/reana_admin.py
+++ b/reana_server/reana_admin.py
@@ -28,6 +28,7 @@ from reana_db.models import (
     AuditLogAction,
     QuotaHealth,
     Resource,
+    ResourceType,
     User,
     UserResource,
     UserTokenStatus,
@@ -521,13 +522,14 @@ def list_quota_resources(ctx, admin_access_token):
     "--email",
     "emails",
     multiple=True,
+    required=True,
     help=(
         "The emails of the users. "
         "E.g. --email johndoe@example.org --email janedoe@example.org"
     ),
 )
 @click.option(
-    "--resource", "-r", help="Specify quota resource. e.g. cpu, disk.", required=True
+    "--resource", "-r", "resource_type", help="Specify quota resource. e.g. cpu, disk.",
 )
 @click.option("--resource-name", "-n", help="Name of resource.")
 @click.option(
@@ -535,30 +537,30 @@ def list_quota_resources(ctx, admin_access_token):
 )
 @admin_access_token_option
 @click.pass_context
-def set_quota_limit(ctx, admin_access_token, emails, resource, resource_name, limit):
+def set_quota_limit(
+    ctx, admin_access_token, emails, resource_type, resource_name, limit
+):
     """Set quota limits to the given users per resource."""
     try:
         for email in emails:
             error_msg = None
+            resource = None
             user = _get_user_by_criteria(None, email)
-            try:
-                resource = Resource.query.filter(
-                    Resource.type_ == resource
-                ).one_or_none()
-            except MultipleResultsFound:
-                # If case of multiple resources of the same type query by the resource_name
-                if resource_name:
-                    resource = Resource.query.filter_by(
-                        name=resource_name
-                    ).one_or_none()
-                else:
-                    error_msg = (
-                        f"ERROR: There are more than one `{resource}` resource. "
-                        "Please provide resource name with `--resource-name` option to specify the exact resource."
+
+            if resource_name:
+                resource = Resource.query.filter_by(name=resource_name).one_or_none()
+            elif resource_type in ResourceType._member_names_:
+                resources = Resource.query.filter_by(type_=resource_type).all()
+                if resources and len(resources) > 1:
+                    click.secho(
+                        f"ERROR: There are more than one `{resource_type}` resource. "
+                        "Please provide resource name with `--resource-name` option to specify the exact resource.",
+                        fg="red",
+                        err=True,
                     )
-            except Exception:
-                Session.rollback()
-                resource = None
+                    sys.exit(1)
+                else:
+                    resource = resources[0]
 
             if not user:
                 error_msg = f"ERROR: Provided user {email} does not exist."
@@ -568,7 +570,7 @@ def set_quota_limit(ctx, admin_access_token, emails, resource, resource_name, li
                     for resource in Resource.query
                 ]
                 error_msg = (
-                    "ERROR: Provided resource does not exist. Available "
+                    f"ERROR: Provided resource `{resource_name or resource_type}` does not exist. Available "
                     f"resources are: {', '.join(resources)}."
                 )
             if error_msg:

```

Reply:
Looks like a nice solution, thanks!